Libplanet/Net/Swarm.cs

```diff
@@ -1472,114 +1468,78 @@ private void BroadcastTxIds(IEnumerable<TxId> txIds)
 
-                // We can't recover with OperationCanceledException and
-                // ObjectDisposedException. so just re-throw them.
-                catch (ObjectDisposedException)
-                {
-                    throw;
-                }
-                catch (OperationCanceledException)
-                {
-                    throw;
-                }
-                catch (Exception e)
-                {
-                    if (retry > 0)
+                    try
                     {
-                        _logger.Error(
-                            e,
-                            $"{nameof(FillBlocksAsync)}() failed; retrying...: {e}"
+                        long currentTipIndex = blockChain.Tip?.Index ?? -1;
+                        long receivedBlockCount = currentTipIndex - previousTipIndex;
+
+                        synced = await FillBlocksAsync(
+                            peer,
+                            blockChain,
+                            stop,
+                            progress,
+                            totalBlockCount,
+                            receivedBlockCount,
+                            evaluateActions,
+                            cancellationToken
                         );
-                        retry--;
+                        break;
                     }
-                    else
+
+                    // We can't recover with OperationCanceledException and
+                    // ObjectDisposedException. so just re-throw them.
+                    catch (ObjectDisposedException)
+                    {
+                        throw;
+                    }
+                    catch (OperationCanceledException)
+                    {
+                        throw;
+                    }
+                    catch (Exception e)
                     {
-                        if (blockChain.Id != synced.Id)
+                        if (retry > 0)
                         {
-                            blockChain.Store.DeleteChainId(synced.Id);
+                            _logger.Error(
+                                e,
+                                $"{nameof(FillBlocksAsync)}() failed; retrying...: {e}"
+                            );
+                            retry--;
+                        }
+                        else
+                        {
+                            throw;
                         }
-
-                        throw;
                     }
                 }
             }
+            finally
+            {
+                if (synced is BlockChain<T> syncedNotNull &&
+                    !syncedNotNull.Id.Equals(blockChain?.Id))
+                {
+                    blockChain.Swap(synced, evaluateActions);
+                }
 
-            return synced;
+                IStore store = BlockChain.Store;
+                foreach (Guid chainId in store.ListChainIds().ToList())
```

Comment:
Does `.ToList()` here purpose to consume the generator ahead of time?

Reply:
Yes. more precisely, it prevents `InvalidOperationException` because `store.DeleteChainId` effects `sotre.ListChainIds()`'s result.